btrdbextras/dq.py

```diff
@@ -0,0 +1,409 @@
+import re
+import warnings
+from tabulate import tabulate
+
+import btrdb
+from btrdb.stream import StreamSet, Stream
+from btrdb.utils.general import pointwidth
+from btrdb.utils.timez import ns_delta, to_nanoseconds
+
+KNOWN_DISTILLER_TYPES = ["repeats", "duplicate-times", "zeros"]
+
+class AmbiguousDistillateError(ValueError):
+    """
+    Raised when a distillate stream's name contains references to multiple known
+    data quality distiller names
+    """
+    pass
+
+class DQDistillate(Stream):
+    """
+    Subsets a Stream object and allows for identfication of data quality events
+
+    Parameters
+    ----------
+    btrdb : BTrDB
+        A reference to the BTrDB object connecting this stream back to the
+		physical server.
+    uuid : UUID
+        The unique UUID identifier for this stream.
+    """
+    def __init__(self, btrdb, uu):
+        # gives all same attrs/methods as Stream
+        super().__init__(btrdb, uu)
+
+        # NOTE: this involves determining distiller type based on the distillate 
+        # stream name, so we will need to be careful how we name distillates
+        types = re.findall(r"(?=("+'|'.join(KNOWN_DISTILLER_TYPES)+r"))", self.name)
+        if len(types) == 0:
+            raise ValueError(f"unknown distiller type. Must be one of [{', '.join(KNOWN_DISTILLER_TYPES)}]")
+        if len(types) > 1:
+            raise AmbiguousDistillateError(f"ambiguous distiller name. contains references to [{', '.join(types)}]")
+        self.type = types[0]
+
+    def contains_event(self, start=None, end=None, depth=30):
+        """
+        Tells whether a distillate stream contains an event, which is denoted by 1 values
+
+        Parameters
+        ----------
+        start: (optional) datetime, datetime64, float, str
+            start time of period to search for events
+        end: (optional) datetime, datetime64, float, str
+            end time of period to search for events
+        depth: (optional) int
+            The precision of the window duration as a power of 2 in nanoseconds.
+            e.g 30 would make the window duration accurate to roughly 1 second
+        
+        Returns
+        -------
+        bool
+            Returns bool indicating whether or not the distillate stream contains an event
+        """
+        start = to_nanoseconds(start) or self.earliest()[0].time
+        # adding 1 to end time because end is exclusive in windows()
+        end = to_nanoseconds(end) or self.latest()[0].time + 1
+        # There's no event if there's no data
+        if start is None and end is None:
+            return False
+        width = end - start
+        windows, _ = zip(*self.windows(start, end, width, depth))
+        return any(w.max >= 1 for w in windows)
```

Comment:
It's interesting to iterate over this, since there should always be only 1 window in the list (since `width` = `end` - `start`. Or, is that not always true?

Reply:
yea... I think that's right. I was iterating originally because I was doing width a different way before, but I think you're right, we won't need to

Reply:
I do wonder if the user could mess with the depth parameter and cause it to return more than one window? I'm not really sure, but I'm tempted to keep it as is because I can't confidently say that it's impossible

Reply:
I'm not sure but that's a good point. If it seems hard to look closer & make any guarantees, then I'd be okay with merging as-is.